File: src/auth/jwt.js
// JWT + role-based access, from scratch.
//
// A JWT is just three base64url strings glued with dots:
//   header.payload.signature
// header  = {"alg":"HS256","typ":"JWT"}       (how it's signed)
// payload = {"sub":"...","role":"admin","exp":123}  (the claims)
// signature = HMAC-SHA256(header + "." + payload, secret)
//
// The server hands one out at login and then TRUSTS ITS OWN SIGNATURE
// instead of a session database: if the signature checks out, the claims
// inside (who you are, what ROLE you have, when it expires) must be exactly
// what we signed — any tampering breaks the HMAC. That's the whole trick,
// and it's why this file has zero dependencies: it's one HMAC, same as the
// Stripe webhook check.
//
// Roles here: "admin" can do everything (billing simulate, all admin data);
// "viewer" is read-only (dashboard data, no mutations).

import { createHmac, timingSafeEqual } from "node:crypto";

const b64url = (buf) => Buffer.from(buf).toString("base64url");
const fromB64url = (s) => Buffer.from(s, "base64url").toString();

export function sign(payload, secret, { expiresInSec = 7200, now = () => Date.now() } = {}) {
  const header = b64url(JSON.stringify({ alg: "HS256", typ: "JWT" }));
  const body = b64url(
    JSON.stringify({ ...payload, iat: Math.floor(now() / 1000), exp: Math.floor(now() / 1000) + expiresInSec })
  );
  const sig = createHmac("sha256", secret).update(`${header}.${body}`).digest("base64url");
  return `${header}.${body}.${sig}`;
}

export function verify(token, secret, { now = () => Date.now() } = {}) {
  const parts = String(token ?? "").split(".");
  if (parts.length !== 3) return { valid: false, why: "malformed" };

  const [header, body, sig] = parts;
  const expected = createHmac("sha256", secret).update(`${header}.${body}`).digest("base64url");
  const a = Buffer.from(expected);
  const b = Buffer.from(sig);
  if (a.length !== b.length || !timingSafeEqual(a, b)) {
    return { valid: false, why: "bad_signature" }; // tampered or wrong secret
  }

  let payload;
  try {
    payload = JSON.parse(fromB64url(body));
  } catch {
    return { valid: false, why: "malformed" };
  }
  if (payload.exp && payload.exp <= Math.floor(now() / 1000)) {
    return { valid: false, why: "expired" }; // a stolen old token is useless
  }
  return { valid: true, payload };
}

// createAuth — the login + gate pair a server actually uses.
// users: { "email": { password, role } }   (demo store; hash passwords in prod)
export function createAuth({ users, secret, expiresInSec = 7200, now = () => Date.now() } = {}) {
  if (!secret) throw new Error("createAuth needs a secret"); // no silent insecurity
  if (!users || Object.keys(users).length === 0) throw new Error("createAuth needs users");

  // Exchange credentials for a signed token. Same vague error for wrong
  // email vs wrong password — don't help attackers enumerate accounts.
  function login(email, password) {
    const user = users[email];
    if (!user || user.password !== password) return null;
    return {
      token: sign({ sub: email, role: user.role }, secret, { expiresInSec, now }),
      role: user.role,
      expiresInSec,
    };
  }

  // The gate: reads "Authorization: Bearer <jwt>", verifies, checks role.
  // Returns the claims when allowed, null when not — caller sends the 401/403.
  function guard(req, allowedRoles) {
    const header = req.headers?.authorization ?? "";
    const token = header.startsWith("Bearer ") ? header.slice(7) : null;
    const result = verify(token, secret, { now });
    if (!result.valid) return null;
    if (allowedRoles && !allowedRoles.includes(result.payload.role)) return null; // wrong role
    return result.payload;
  }

  return { login, guard };
}

File: internal/authjwt/jwt.go
// Package authjwt implements JSON Web Tokens and role-based access from
// scratch, with nothing but crypto/hmac.
//
// A JWT is three base64url strings glued with dots:
//
//	header.payload.signature
//	header    {"alg":"HS256","typ":"JWT"}
//	payload   {"sub":"...","role":"admin","exp":...}
//	signature HMAC-SHA256(header + "." + payload, secret)
//
// The server hands one out at login and then trusts its own signature instead
// of keeping a session database: if the HMAC verifies, the claims inside must
// be exactly what we signed, because any tampering changes the digest. That is
// the entire trick, and it is why this file has no dependencies.
//
// Roles: "admin" can do everything, "viewer" is read-only, "user" is an org
// member whose visibility is resolved per request from the org directory.
package authjwt

import (
	"crypto/hmac"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/base64"
	"encoding/json"
	"errors"
	"strings"
	"time"
)

var (
	ErrMalformed    = errors.New("jwt: malformed")
	ErrBadSignature = errors.New("jwt: bad signature")
	ErrExpired      = errors.New("jwt: expired")
)

// Claims is the payload we sign. Extra fields are not carried: everything the
// gateway authorises on is here.
type Claims struct {
	Sub  string `json:"sub"`
	Role string `json:"role"`
	IAT  int64  `json:"iat"`
	Exp  int64  `json:"exp"`
}

func b64(b []byte) string            { return base64.RawURLEncoding.EncodeToString(b) }
func unb64(s string) ([]byte, error) { return base64.RawURLEncoding.DecodeString(s) }

// Sign mints a token valid for expiresIn.
func Sign(sub, role, secret string, expiresIn time.Duration, now func() time.Time) string {
	if now == nil {
		now = time.Now
	}
	t := now()
	header := b64([]byte(`{"alg":"HS256","typ":"JWT"}`))
	payload, _ := json.Marshal(Claims{
		Sub:  sub,
		Role: role,
		IAT:  t.Unix(),
		Exp:  t.Add(expiresIn).Unix(),
	})
	body := b64(payload)
	return header + "." + body + "." + sign(header+"."+body, secret)
}

func sign(signingInput, secret string) string {
	m := hmac.New(sha256.New, []byte(secret))
	m.Write([]byte(signingInput))
	return b64(m.Sum(nil))
}

// Verify checks the signature and the expiry, in that order.
func Verify(token, secret string, now func() time.Time) (Claims, error) {
	if now == nil {
		now = time.Now
	}
	parts := strings.Split(token, ".")
	if len(parts) != 3 {
		return Claims{}, ErrMalformed
	}
	expected := sign(parts[0]+"."+parts[1], secret)

	// Constant-time compare: a byte-by-byte early exit leaks, through timing,
	// how much of a forged signature was correct.
	if subtle.ConstantTimeCompare([]byte(expected), []byte(parts[2])) != 1 {
		return Claims{}, ErrBadSignature
	}

	raw, err := unb64(parts[1])
	if err != nil {
		return Claims{}, ErrMalformed
	}
	var c Claims
	if err := json.Unmarshal(raw, &c); err != nil {
		return Claims{}, ErrMalformed
	}
	if c.Exp != 0 && c.Exp <= now().Unix() {
		return Claims{}, ErrExpired // a stolen old token is useless
	}
	return c, nil
}

// ---- the login + gate pair a server actually uses --------------------------

type User struct {
	Password string
	Role     string
}

type Auth struct {
	users     map[string]User
	secret    string
	expiresIn time.Duration
	now       func() time.Time
}

type Session struct {
	Token        string `json:"token"`
	Role         string `json:"role"`
	ExpiresInSec int    `json:"expiresInSec"`
}

func New(users map[string]User, secret string, expiresIn time.Duration, now func() time.Time) (*Auth, error) {
	if secret == "" {
		return nil, errors.New("auth needs a secret") // no silent insecurity
	}
	if len(users) == 0 {
		return nil, errors.New("auth needs users")
	}
	if expiresIn == 0 {
		expiresIn = 2 * time.Hour
	}
	return &Auth{users: users, secret: secret, expiresIn: expiresIn, now: now}, nil
}

// Login exchanges credentials for a signed token. Wrong email and wrong
// password fail identically, so an attacker cannot enumerate accounts.
func (a *Auth) Login(email, password string) *Session {
	u, ok := a.users[email]
	if !ok || subtle.ConstantTimeCompare([]byte(u.Password), []byte(password)) != 1 {
		return nil
	}
	return &Session{
		Token:        Sign(email, u.Role, a.secret, a.expiresIn, a.now),
		Role:         u.Role,
		ExpiresInSec: int(a.expiresIn.Seconds()),
	}
}

// Guard reads "Authorization: Bearer <jwt>", verifies it, and checks the role.
// It returns the claims when allowed and nil when not; the caller sends the 401
// so it can shape the error body.
func (a *Auth) Guard(authHeader string, allowedRoles ...string) *Claims {
	token := strings.TrimPrefix(authHeader, "Bearer ")
	if token == authHeader { // prefix was absent
		return nil
	}
	return a.GuardToken(token, allowedRoles...)
}

// GuardToken is the same check for wires that cannot set headers — the SSE
// stream and the WebSocket upgrade both carry the token in the query string.
func (a *Auth) GuardToken(token string, allowedRoles ...string) *Claims {
	c, err := Verify(token, a.secret, a.now)
	if err != nil {
		return nil
	}
	if len(allowedRoles) > 0 {
		for _, r := range allowedRoles {
			if c.Role == r {
				return &c
			}
		}
		return nil // wrong role
	}
	return &c
}

func (a *Auth) Secret() string           { return a.secret }
func (a *Auth) ExpiresIn() time.Duration { return a.expiresIn }
